python/phase-2.md

```diff
@@ -0,0 +1,996 @@
+  using `functools.wraps`), we could have it return something like `ActivityDefinition[ParamTypes, ReturnType]` (which
+  does not get the transitive benefits). The reason we might want this is then we can have `@activity.method_defn` or
+  similar that is smart enough to convert to an activity definition. I have been trying to avoid actually wrapping
+  functions so far.
+* We could support the following with an overload:
+
+    ```python
+    workflow.execute_activity(ActivityClass.some_activity, "foo", method=True)
+    ```
+  This would work with a special overload like:
+
+    ```python
+    @overload
+    async def execute_activity(
+        activity: Callable[[Any, LocalParamType], LocalReturnType],
+        arg: LocalParamType,
+        /,
+        *,
+        method: Literal[True],
+    ```
+  This seems like a fairly sane way, but requires callers pass a special parameter to call
+* Another outstanding question - is there added benefit _at this time_ to support registering a class full of activity
+  methods instead of individually? I am thinking not (yet).
+
+### Workflow Contracts
+
+Sometimes users may want to separate the workflow into a contract and an impl so that they can be separated. We will
+support this approach.
+
+Here is what it looks like with `typing.Protocol`s:
+
+```python
+@workflow.defn
+class MyWorkflow(Protocol):
+    @workflow.run
+    def run(self, arg: str) -> List[str]:
+        ...
+
+@workflow.defn(name=MyWorkflow.__name__)
+class MyWorkflowImpl:
+    @workflow.run
+    def run(self, arg: str) -> List[str]:
+        return [arg]
+
+async def run_workflow(client: Client) -> List[str]:
+    return await client.execute_workflow(MyWorkflow.run)
+
+def create_worker(client: Client) -> Worker:
+    return Worker(workflows=[MyWorkflowImpl])
+```
+
+And here's what it looks like with abstract classes:
+
+```python
+@workflow.defn
+class MyWorkflow(ABC):
+  @workflow.run
+  @abstractmethod
+  def run(self, arg: str) -> List[str]:
+    raise NotImplementedError
+
+@workflow.defn(name=MyWorkflow.__name__)
+class MyWorkflowImpl(MyWorkflow):
+    @workflow.run
+    def run(self, arg: str) -> List[str]:
+        return [arg]
+
+async def run_workflow(client: Client) -> List[str]:
+    return await client.execute_workflow(MyWorkflow.run)
+
+def create_worker(client: Client) -> Worker:
+    return Worker(workflows=[MyWorkflowImpl])
+```
+
+Notes:
+
+* This has no special logic to handle this approach. The same approach can be used to separate activity class method
+  contracts from their impls.
+* It is unfortunate that we require a `@workflow.defn` on both the base and impl classes.
+  * We want to require this decorator at client-side _and_ worker-side to ensure name sharing
+  * With `Protocol` it is obvious why, there is technically no code shared
+  * With abstract base classes, we maybe could support `@workflow.defn(extends=MyWorkflow)` but that doesn't give you
+    much
```

Comment:
Why does it not give you much? It should be able to set the name correctly, yes? If so, you could do 

`@workflow.defn(implements=MyWorkflow)` for protocols and I think that's quite nice.

I have no problem with both places declaring the same decorators. I think it's a feature rather than a downside. It's double checking you're consistent, and it makes it visually obvious without referring to the protocol/base what the class is doing. Super minimal amount of extra typing for some additional clarity.

Reply:
> Why does it not give you much? It should be able to set the name correctly, yes? If so, you could do

It doesn't give you much because you're already extending `MyWorkflow` with the Python syntax. I don't need a custom extension approach too. I can derive it at runtime. The only benefit would be getting the name right, but if we move to proper inherited decorators that benefit wouldn't even exist either.

> If so, you could do `@workflow.defn(implements=MyWorkflow)` for protocols and I think that's quite nice.

This bullet was discussing abstract base classes. Protocols are meant to be separate from the implementation. In Python if you want explicit "implements" you'd use an abstract base class and properly extend. There is no benefit to using a Protocol over an abstract base class if you're already sharing code.

> I have no problem with both places declaring the same decorators

My only problem is getting the name right. The two classes need to have the same name.